### PR TITLE
fix mysql config read host (.my.cnf) #6761

### PR DESCRIPTION
**Pull request by genofire**
let mysqllib choose default values)

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
every mysql_* module

##### ANSIBLE VERSION

```
ansible 2.4.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/genofire/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/sbin/ansible
  python version = 2.7.14 (default, Jan  5 2018, 10:41:29) [GCC 7.2.1 20171224]
```

##### ADDITIONAL INFORMATION
see issue

```

```

### CODE REVIEW

**Comment by genofire** [comment]
@mpdehaan ??

**Comment by genofire** [comment]
Hey whats up @Jmainguy and @banyek - a new ansible version, where i have to fix this issue again (on my daily ansible)

**Comment by seanorama** [comment]
Any plans to accept this in? Seems odd that ~/.my.cnf is not being respected.

**Comment by Xyon** [comment]
How does this behave if I don't supply a my.cnf file?

**Comment by genofire** [comment]
It will use `/etc/mysql/my.cnf` - https://dev.mysql.com/doc/refman/5.7/en/option-files.html
Afterwards there are default values (which depence on distro/build)

**Comment by genofire** [comment]
okay, i will not rebase it - if nobody review it ... after a year give some respect to contributer

**Comment by bmalynovytch** [comment]
Dear @genofire,
There has been nearly no code review on MySQL/MariaDB modules for a long time.
A dedicated working group (https://github.com/ansible/community/wiki/MySQL) was setup a couple of weeks ago, made of contributors like you.
We do our best to make things change, but it takes some time.
I don't have much time right now but will do my best to review PRs ASAP, although, I can't review / validate if there are some conflicts or CI failures.
Rebasing your code on my own requires too much effort (I'd have to do it on many PRs, as large changes occurred recently on `mysql_user` module). If you want those changes to be pushed soon and even backported in 2.7, you'll need to rebase and fix conflicts again, sorry :(

**Comment by genofire** [comment]
I will, after i get feedback, what should a change also ...

**Comment by bmalynovytch** [comment]
@genofire your changes seem reasonable and, I suppose, will fix the issue (#6761), but I won't review if it isn't rebased, as I can't test it on devel branch as is.

**Comment by genofire** [comment]
Good, then i does not need to test it ...
Hope it does not take another year

**Comment by bmalynovytch** [comment]
@genofire you'll need to change https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/doc_fragments/mysql.py#L20 to fix the documentation, to match your new default value for `login_host`

**Comment by genofire** [comment]
@bmalynovytch i could not define e default value -> is it okay this way?

File: lib/ansible/plugins/doc_fragments/mysql.py
Comment (bmalynovytch) [inline lib/ansible/plugins/doc_fragments/mysql.py]:
```suggestion
      - Host running the database. Default inherited from client library
```

**Comment by bmalynovytch** [comment]
Appart from the recommendation on the previous comment, your changes seem fine.
I'll do some tests to ensure it doesn't break anything.

File: lib/ansible/plugins/doc_fragments/mysql.py
Comment (genofire) [reply lib/ansible/plugins/doc_fragments/mysql.py]: Wow, i did not know github has this function ;)

Comment (bmalynovytch) [reply lib/ansible/plugins/doc_fragments/mysql.py]: it's quite new :)
The button to use is the document icon with `+/-` symbol, on the right of the `Preview` tab in the comment box

**Comment by genofire** [comment]
And now? after a month? - do not get it unmergable again

**Review by bmalynovytch — APPROVED** [review]

**Comment by bmalynovytch** [comment]
shipit

**Comment by pratikgadiya12** [comment]
@genofire 
 I tested the patch and it does work for host however, I think we need to modify the PR for other variables as well that are available in ~/.my.cnf (like login_password, login_user,..)

Reference:
[1]. https://docs.ansible.com/ansible/latest/modules/mysql_user_module.html#notes
[2]. https://docs.ansible.com/ansible/2.6/modules/mysql_user_module.html?highlight=mysql_user

CC: @bmalynovytch

**Comment by genofire** [comment]
`login_password`, `login_user` should already work ... only only could discuss over `login_port` ... maybe on next PR / issue ?

**Comment by bmalynovytch** [comment]
> `login_password`, `login_user` should already work ... only only could discuss over `login_port` ... maybe on next PR / issue ?

Completely agree with you (see https://github.com/ansible/ansible/issues/43424#issuecomment-496099975)

**Comment by bmalynovytch** [comment]
bot_status

**Comment by pratikgadiya12** [comment]
shipit

**Comment by stintel** [comment]
Unfortunately this will break all mysql_ tasks where login_host is not set. I'm pretty sure this will upset a lot of people. While I do like to see this problem solved, the suggested solution is imo not the best way forward.

The problem with at least PyMySQL is that if the Connection class is instantiated with any of the following parameters, the library will simply use them instead or the values set in the config file:
 - user
 - password
 - host
 - database
 - socket
 - port
 - bind-address
 - default-character-set
 - ssl-{ca,capath,cert,key,cipher}

Instead of potentially breaking loads of playbooks out there and pissing off users, I propose the following solution:

Add a new boolean parameter prefer_config_file with default value false. Setting the parameter to true would unset the defaults for login_host and login_port. This way, existing playbooks where user rely on the default login_host being localhost and the default login_port being 3306 would continue to work. We can then solve our problem by setting prefer_config_file to true. The only downside is that we would always need to set host and port in the .my.cnf file, even if we're running on localhost:3306.

**Comment by genofire** [comment]
Is there any reason, why this is not merged?

**Comment by bmalynovytch** [comment]
@genofire just ask the bot

bot_status

**Comment by bmalynovytch** [comment]
@genofire we miss another shipit from maintainers or others from community

**Comment by stintel** [comment]
shipit

**Comment by stintel** [comment]
I guess I don't quality as "others from community"?

**Comment by Andersson007** [comment]
If i have time tomorrow, i’ll look at this

**Comment by bmalynovytch** [comment]
@stintel you should be identified as member of community, not maintainer

**Comment by Andersson007** [comment]
Could you please add the integration tests for this case before my review, or point some existing out that indicate that nothing was broken down

**Comment by banyek** [comment]
yeah I abandoned long time ago, but I'd like to come back if I may

**Review by Andersson007 — CHANGES_REQUESTED** [review]
@genofire , waiting for your response about the integration tests, see my previous comment

**Comment by genofire** [comment]
@Andersson007 I do not know how ...

**Comment by Andersson007** [comment]
@genofire , hey, sorry for late reply, was on holiday.
integration tests are simple playbooks, look at files in test/integration/targets/mysql_*/tasks/

You can run tests locally before pushing to github, from Ansible source tree directory, e.g. (docker must be installed) and you need to run ```source hacking/env-setup``` before
```
ansible-test integration mysql_db --docker ubuntu1804 -vvvv > ~/test.log
```
change mysql_db to subdirectory name, e.g. mysql_user, etc., it depends on where your test file is. you can also choose another distribution, like centos7 for example
if errors occur you can check the log content and fix

**Comment by genofire** [comment]
hey, sorry - i am out, not willing to maintain this PR anymore.
> i unsubscribe the notification and allow maintainers to change this edit

so feel free to fix it and merge it or just to close

**Comment by Andersson007** [comment]
close_me

**Pull request closed without merging**
